nixos/knot: allow full configuration by nix values

vcunat opened the pull request:
This seems desired by multiple people.  Let's work out how to do this, starting with this prototype.  See the commit messages.

`settings` RFC for reference: https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md

vcunat committed:
nixos/knot: allow full configuration by nix values (RFC 42)

vcunat commented:
One idea not yet implemented:
```nix
{
       # Example usage:
       zones.foo = "inline string";
       zones.bar = ./some-file;
       zones.baz = some derivation; # e.g. writeText, fetch* or whatever

       # where implementation might be something like:
       nixConfig.zone = flip mapAttrs zones
            (n: v: {
              storage = "/var/lib/knot";
              file = if isString v then pkgs.writeText n v else v;
            });
}
```

vcunat commented:
So far I absolutely haven't looked at aspects like what happens on errors.  Some common problems might lead to horrible nix traces, for example.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/generating-yaml-in-a-nix-module/2519/5

mweinelt commented:
> One idea not yet implemented:
> 
> ```nix
> {
>        # Example usage:
>        zones.foo = "inline string";
>        zones.bar = ./some-file;
>        zones.baz = some derivation; # e.g. writeText, fetch* or whatever
> 
>        # where implementation might be something like:
>        nixConfig.zone = flip mapAttrs zones
>             (n: v: {
>               storage = "/var/lib/knot";
>               file = if isString v then pkgs.writeText n v else v;
>             });
> }
> ```

Ultimately something like this would be awesome https://github.com/kirelagin/dns.nix.

vcunat commented:
Generating the zone files themselves seems like an orthogonal task.  In particular, it should be the same for all DNS servers.  I think I've seen multiple proposals for some such generator already.

sinavir commented:
Hello, any updates on this ? How can I help ?

vcunat commented:
Test and review, I think.  I got basically no reactions here, so I didn't feel like going forward with it.  IIRC that's the only reason why this got left behind by me.

I could rebase a look about possible silent conflicts, but I expect an issue is only in the tests even though it's been years.

mweinelt committed:
nixosTests.knot: use settings format

mweinelt committed:
nixosTests.kea: use knot.settings for configuration

mweinelt commented:
Rebased.

Made the new declarative configuration into the `settings` option and removed `extraConfig`, because merging isn't really feasible.

Updated the remaining knot configs in our tests accordingly.

dummy987654321 commented:
I think we should keep `extraConfig` with a deprecation warning for two reason:
1. This will help users making the transition
2. There may be some setups that are not doable with this pr that we didn't expected

dummy987654321 commented:
Ok looked a bit deeper in knot docs and I believe that the current PR cannot handle modules (https://www.knot-dns.cz/docs/latest/html/modules.html).

It seems to me that all modules are parsed by `sec_list_fa "id"`. Thus it can be handled by iterating into `settings` top-level ang extracting the keys that begin by `mod-` .

dummy987654321 commented:
Here is a patch :
```
diff --git a/nixos/modules/services/networking/knot.nix b/nixos/modules/services/networking/knot.nix
index bea980066cd..bf4186f6c31 100644
--- a/nixos/modules/services/networking/knot.nix
+++ b/nixos/modules/services/networking/knot.nix
@@ -11,6 +11,7 @@ let
         # We output the config section in the upstream-mandated order.
         # Ordering is important due to forward-references not being allowed.
         [ (sec_list_fa "id" nix_def "module") ]
+        ++ map (sec_list_fa "id" nix_def) (filter (hasPrefix "mod-") (attrNames nix_def))
         ++ map (sec_plain nix_def)
           [ "server" "control" ]
         ++ [ (sec_list_fa "target" nix_def "log") ]
```

mweinelt commented:
We're also going to run into ordering problems with mods. 

![image](https://github.com/NixOS/nixpkgs/assets/131599/e697a2a7-6027-42ac-8712-ff2b60a4b4e6)

https://www.knot-dns.cz/docs/latest/html/configuration.html#query-modules

dummy987654321 commented:
Hmm, do you think it is worth continuing that way ? Or should the module let the user provide the ordering ?

mweinelt commented:
We could either make this a fixed order for now, or work with priorities (e.g. nginx location blocks would be the prior art), or maybe even toposort.

vcunat commented:
AFAIK the only unresolved question is whether still allow something like `extraConfig`, even if just for the option of using old style instead of `.settings`?

And possibly, I'm not sure if we could do some improvement around `.keyFiles` or if there's some other practical situation that could break `knotc conf-check`.

vcunat commented:
I didn't have any plan.  Just append it and it's the user's responsibility that it makes sense 🤷🏽

mweinelt commented:
Yes, symlinking it sounds good to me.

> Actually I don't find it ideal to call the service "knot", but that's what we historically have in NixOS.

```
# dpkg -S knot.service
knot: /lib/systemd/system/knot.service
```

That's also the reality in the upstream maintained apt packages.

vcunat commented:
Ah right, let me drop the naming topic in this PR.  It's just... long-term complication of some people getting confused because of using plain "knot" word in different senses.  For example, Turris users got used to that for `knot-resolver` service, by shortening the official name.

dummy987654321 commented:
> Actually I don't find it ideal to call the service "knot", but that's what we historically have in NixOS.

We can rename it. There are functions in nixpkgs specially tailored to do this iirc

vcunat commented:
I know, but after more thought it's probably not worth doing at this point.

dummy987654321 commented:
I also think you should keep `extraConfig`. To achieve this we can make `settings` and `extraConfig` exclusive with an assertion.

vcunat committed:
nixos/knot: also allow config by YAML file

vcunat commented:
I think the PR now has everything that's been suggested.  Also /cc @lheckemann as potentially interested in reacting to this PR.

vcunat commented:
Nit: maybe `assertions` instead of `assert` would give more consistent errors.  At the moment I didn't feel like moving the check away from its place, but it certainly wouldn't be a big deal.

mweinelt commented:
Feel free to improve on that nit, I'm very much inclined to include this in NixOS 23.11. Even in its current state, it is a big improvement.

vcunat committed:
nixos/knot: add release notes and partial compatibility

vcunat commented:
- cleaned up history where I didn't consider it worth keeping the in-between states
- added release notes and partial compatibility with NixOS 23.05 config

I think this is the last call.  There has been too much opportunity for feedback already.

lheckemann reviewed (approved):
:heart:

mweinelt commented:
I think all interested parties have approved. Thanks for your work on this, it makes using knot much more pleasant than working with the upstream config format.

SuperSandro2000 commented:
Naive question: wouldn't this be a bit simpler if knot's yaml parser would be a full yaml 1.1 parser and not only supporting a tiny subset?

SuperSandro2000 commented on nixos/modules/services/networking/knot.nix:
```diff
@@ -5,10 +5,110 @@ with lib;
+  in result;
+
+  configFile = if cfg.settingsFile != null then
+    assert cfg.settings == {} && cfg.keyFiles == [];
```
We should move this to a full fledged module system assertion to make it UX wise a lot better.
Right now I receive the following error which is a bit confusing.

```
       … while evaluating a branch condition

         at /nix/store/4h9mf2ydqjgjawihf3zkshkznm4mkavg-source/lib/customisation.nix:256:8:

          255|       drv' = (lib.head outputsList).value;
          256|     in if drv == null then null else
             |        ^
          257|       lib.deepSeq drv' drv';

       … while calling the 'head' builtin

         at /nix/store/4h9mf2ydqjgjawihf3zkshkznm4mkavg-source/lib/attrsets.nix:820:11:

          819|         || pred here (elemAt values 1) (head values) then
          820|           head values
             |           ^
          821|         else

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: assertion '(((cfg).settings == { }) && ((cfg).keyFiles == [ ]))' failed

       at /nix/store/4h9mf2ydqjgjawihf3zkshkznm4mkavg-source/nixos/modules/services/networking/knot.nix:106:5:

          105|   configFile = if cfg.settingsFile != null then
          106|     assert cfg.settings == {} && cfg.keyFiles == [];
             |     ^
          107|     cfg.settingsFile
```

SuperSandro2000 replied:
Also this isn't mentioned in the release-notes, so I am currently a bit confused how to migrate of the following:

```nix
      keyFiles = [
        config.sops.secrets."knot/keyFile".path
      ];
```

vcunat replied:
If you switch to `settings`, you can keep `keyFiles`.

As for the module, one way is to make the assertion better, another way is to implement full compatibility.

SuperSandro2000 replied:
I ended up with this diff to have an identical knot.conf file without rewriting my config.

```diff
diff --git machines/neon/configuration.nix machines/neon/configuration.nix
index 947b152..594c898 100644
--- machines/neon/configuration.nix
+++ machines/neon/configuration.nix
@@ -134,10 +134,8 @@

     knot = {
       enable = true;
-      keyFiles = [
-        config.sops.secrets."knot/keyFile".path
-      ];
-      extraConfig = /* yaml */ ''
+      settingsFile = pkgs.writeText "knot.conf" /* yaml */ ''
+        include: ${config.sops.secrets."knot/keyFile".path}
         acl:
           - id: transfer
             key: transfer
```

cpu replied:
@SuperSandro2000 Thank you, this was helpful. My configuration looked very similar to yours.

vcunat replied:
Actually, people may want to revert similar config changes when updating past PR https://github.com/NixOS/nixpkgs/pull/258361 - as such includes will surely often fail build-time validation that was disabled by accident.  (I'm sure people do want validation when *not* using secrets.)

SuperSandro2000 commented on nixos/doc/manual/release-notes/rl-2311.section.md:
```diff
@@ -97,6 +97,8 @@
 
 - `pass` now does not contain `password-store.el`.  Users should get `password-store.el` from Emacs lisp package set `emacs.pkgs.password-store`.
 
+- `services.knot` now supports `.settings` from RFC42.  The change is not 100% compatible with the previous `.extraConfig`.
```
This should relay contain what is not compatible or some link to maybe here or somewhere else because right now lonely from this line I am totally unsure what to other than just give it a shot.

cpu replied:
I would also appreciate seeing more detail here. The error generated without making any changes to my previous config was (at least to my eyes) fairly cryptic:

<details>
<summary>Full error trace:</summary>

```
error:
       … while calling the 'head' builtin

         at /nix/store/8lg5579v1k5mc63ypzfagrcl8l32x3bh-source/lib/attrsets.nix:820:11:

          819|         || pred here (elemAt values 1) (head values) then
          820|           head values
             |           ^
          821|         else

       … while evaluating the attribute 'value'

         at /nix/store/8lg5579v1k5mc63ypzfagrcl8l32x3bh-source/lib/modules.nix:807:9:

          806|     in warnDeprecation opt //
          807|       { value = builtins.addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          808|         inherit (res.defsFinal') highestPrio;

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: assertion '(((cfg).settings == { }) && ((cfg).keyFiles == [ ]))' failed

       at /nix/store/8lg5579v1k5mc63ypzfagrcl8l32x3bh-source/nixos/modules/services/networking/knot.nix:106:5:

          105|   configFile = if cfg.settingsFile != null then
          106|     assert cfg.settings == {} && cfg.keyFiles == [];
             |     ^
          107|     cfg.settingsFile
 ```
 </details>

vcunat replied:
The error should be gone after merging PR https://github.com/NixOS/nixpkgs/pull/258361

SuperSandro2000 commented:
> Feel free to improve on that nit, I'm very much inclined to include this in NixOS 23.11 either way. Even in its current state, it is a big improvement.

I don't think it is that much currently. I mostly configured my knot by copying the yaml from the documentation page section by section. Now I would need to translate it to nix code and make sure that the limited parser still understands what I am doing. It would mean I could built my own functions to generate repetitive sections but there are currently no value adding options like nginx' enableACME or complete location blocks like nextcloud or mastodon has.

vcunat commented:
I see a part the other way than you: translating to nix code will _work around_ the limited YAML parser, e.g. it will ensure that ordering of the sections is correct.  Either way, I don't think `settingsFile` will be going away.

SuperSandro2000 commented:
That is an interesting fact I didn't know about. How about we write this into the release notes to motivate people to migrate their config?

vcunat commented:
I think it's OK if they don't migrate.  And `keyFiles` just insert `include: ${file}` lines, so you could put it into your YAML by hand.  I could figure out how to layer this with the compat helper `mkChangedOptionModule`.

vcunat commented:
Solving ordering was a necessity in this PR.  Nix attrsets are unordered, so the user no longer can reliably affect ordering.

vcunat commented:
Uh, so I guess continuing on PR https://github.com/NixOS/nixpkgs/pull/258361

ju1m commented on nixos/modules/services/networking/knot.nix:
```diff
@@ -60,11 +160,21 @@ in {
-        type = types.lines;
-        default = "";
+      settings = mkOption {
+        type = types.attrs;
```
Thanks for this huge improvement! Please be aware that this `type.attrs` does not merge the `settings` as expected with other options, especially when the configuation is split between modules. Using a freeform submodule as suggested in the [RFC followed by this PR](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md#part-1-1) would solve that.

vcunat replied:
That should be in PR #304373
